### PR TITLE
scramble password at search time

### PR DESCRIPTION
**Pull request by alexmasselot**

### CODE REVIEW

File: components/api/flaskr/passwordChecker/substitute.py [hunk 1]
```diff
@@ -57,167 +55,252 @@ def __init__(self,
...
+        ['paf']
+        >>> CharacterProjector().index_password('hello')
+        ['heiio']
+        >>> CharacterProjector().index_password('he6a')
+        ['heba', 'hega']
+        """
+        cleaned_up = self.trailing_cleaning_regexp.sub('', password)
+        return self.project_word(cleaned_up)
+
+    def potential_indexes(self, password: str):
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/substitute.py:100]: better named potential_indexed

File: components/api/flaskr/passwordChecker/substitute.py [hunk 2]
```diff
@@ -1,35 +1,33 @@
...
 
 
-# if a word such as 'bonjour' is passed, we want to generate variants, by substituting letters, adding punctuation and
-# so on. This file contains the function to generate al those variants
-
-class SubstitutionParams:
-    substitute_characters: int
-    append_punctuation: bool
-    append_number: bool
+# if a word such as 'bonjour' is passed, we want to check variants, by substituting letters, adding punctuation and
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/substitute.py:5]: that would deserve more comments

File: components/api/flaskr/passwordChecker/dictionary.py [hunk 3]
```diff
@@ -95,8 +83,28 @@ def contains(self, password: str):
...
+            for dictionary in self.dictionaries:
+                if dictionary.contains_exact(to_search):
+                    return True
 
         return False
+
+    def load_dictionary(self, filename: str):
+        """
+        Load all the word (one per line) from a text file into a PasswordDictionary.
+        Words are trimmed and substitution or character, as well as appending punctuation and and number can be generated
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/dictionary.py:97]: This is obsolete

Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/dictionary.py:97]:
```suggestion
        Words are trimmed and substitution or character, as well as appending punctuation and and number can be generated
```

File: components/api/flaskr/passwordChecker/robustness.py [hunk 4]
```diff
@@ -35,6 +35,6 @@ def compute_robustness(password: str, dictionaryChecker: DictionaryChecker):
     """
     return PasswordRobustness(
         length_score=password_length_score(password),
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/robustness.py:37]: rename to compute_password_length

File: components/api/flaskr/passwordChecker/substitute.py [hunk 5]
```diff
@@ -57,167 +55,252 @@ def __init__(self,
...
+
+        if self.trailing_regexp:
+            m = self.trailing_regexp.search(password)
+            if m:
+                truncated_passwords = [password[0:i] for i in range(m.start(), m.end() + 1)]
+                return list(chain(*[self.project_word(p) for p in truncated_passwords]))
+
+        return self.project_word(password)
+
+    def trim(self, word):
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/substitute.py:125]: This method looks useless

File: components/api/flaskr/passwordChecker/substitute.py [hunk 6]
```diff
@@ -57,167 +55,252 @@ def __init__(self,
...
-    return [prefix + subst + suffix for subst in _subDict[char_to_substitute] if subst != char_to_substitute]
+                if self.scrambling_params.trailing_punctuations:
+                    trailing_chars = _numbers
+            return CharacterProjector.build_regexp(trailing_chars, is_tail=True,
+                                                   max_length=self.scrambling_params.max_trailing)
+        return None
+
+    @staticmethod
+    def build_regexp(chars: list, is_not: bool = False, is_head: bool = False, is_tail: bool = False,
+                     max_length: int = 0):
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/substitute.py:215]: what is thismax_length used for

File: components/api/flaskr/passwordChecker/substitute.py [hunk 7]
```diff
@@ -57,167 +55,252 @@ def __init__(self,
...
+        if is_tail:
+            str_regexp += '$'
+        if is_head:
+            str_regexp = '^' + str_regexp
+        return re.compile(str_regexp)
+
+    @staticmethod
+    def build_character_projection(subst_dict: dict):
+        """
+        from a substDict, build the project (the reverse one)
```
Comment (alexmasselot) [inline components/api/flaskr/passwordChecker/substitute.py:277]: make it clearer.
Why is that a static method ? 
subst_dict is a properties

File: components/api/flaskr/app.py [hunk 8]
```diff
@@ -1,30 +1,24 @@
 from flask import Flask, jsonify
```
Comment (mriffe) [inline components/api/flaskr/app.py:1]: Test

**Review by alexmasselot — COMMENTED** [review]
just a few comments

**Pull request merged**